File: 1695-maximum-erasure-value/main.cpp
class Solution {
public:
  int maximumUniqueSubarray(vector<int>& nums) {
    int score = 0;
    unordered_set<int> collection;
    int subscore = 0;
    vector<int>::const_iterator left = nums.begin();
    for (const auto& right: nums) {
      auto found = collection.find(right);
      if (found == collection.end()) {
        collection.insert(right);
      } else {
        for (; *left!=right; left++) {
          subscore -= *left;
          collection.erase(*left);
        }
        subscore -= *left;
        left++;
      }
      subscore += right;
      score = max(score, subscore);
    }
    return score;
  }
};

File: 1695-maximum-erasure-value/main.rs
use std::collections::HashSet;

impl Solution {
    pub fn maximum_unique_subarray(nums: Vec<i32>) -> i32 {
        let mut maximum: i32 = 0;
        let mut inner: i32 = 0;
        let mut set = HashSet::<i32>::new();
        let mut l = 0;
        let mut r = 0;
        while r < nums.len() {
            while l < nums.len() && l <= r {
                if set.contains(&nums[r]) {
                    set.remove(&nums[l]);
                    inner -= nums[l];
                } else {
                    break;
                }
                l += 1;
            }
            set.insert(nums[r]);
            inner += nums[r];
            if maximum < inner {
                maximum = inner;
            }
            r += 1;
        }
        return maximum;
    }
}
